flang/lib/Lower/ConvertVariable.cpp

```diff
@@ -386,9 +386,36 @@ static fir::GlobalOp defineGlobal(Fortran::lower::AbstractConverter &converter,
   mlir::Location loc = converter.genLocation(sym.name());
   bool isConst = sym.attrs().test(Fortran::semantics::Attr::PARAMETER);
   fir::GlobalOp global = builder.getNamedGlobal(globalName);
+  mlir::Type symTy = converter.genType(var);
+
   if (global && globalIsInitialized(global))
     return global;
-  mlir::Type symTy = converter.genType(var);
+  // If this is an array, check to see if we can use a dense attribute
+  // with a tensor mlir type.  This optimization currently only supports
+  // rank-1 Fortran arrays of integer, real, or logical. The tensor
+  // type does not support nested structures which is needed for
+  // complex numbers.
+  // To get multidimensional arrays to work, we will have to use column major
+  // array ordering with the tensor type (so it matches column major ordering
+  // with the Fortran fir.array).  By default, tensor type assume row major
+  // ordering. How to create this tensor type is to be determined.
+  if (symTy.isa<fir::SequenceType>() && sym.Rank() == 1 &&
+      !Fortran::semantics::IsAllocatableOrPointer(sym)) {
+    auto eleTy = symTy.cast<fir::SequenceType>().getEleTy();
+    if (eleTy.dyn_cast<mlir::IntegerType>() ||
+        eleTy.dyn_cast<mlir::FloatType>() ||
+        eleTy.dyn_cast<fir::LogicalType>()) {
+      const auto *details =
+          sym.detailsIf<Fortran::semantics::ObjectEntityDetails>();
+      if (details->init()) {
+        global = Fortran::lower::createDenseGlobal(
```

Comment:
At one point in my development of this feature, this function was changed to create the global first with an empty attribute field. Unfortunately, I was unable to just add the attribute in that case. The global kept coming up with an empty attribute. I will add an assert and if we find this to be a problem down the road, I can revisit it.

Reply:
After making a change, createDenseGlobal() can return 0. So, instead of an assert, make sure global is non-zero, otherwise, call builder.createGloba()l, etc.